File: Reference/tests/Brontide.Reference.Interchange.Tests/Portable/PortableAuthorityAndResourceTests.cs
using Brontide.Reference.Experimental.Binding.Portable;

namespace Brontide.Reference.Interchange.Tests.Portable;

/// <summary>Neutral vectors PB-18 through PB-32: local authority and referenced resources.</summary>
public sealed class PortableAuthorityAndResourceTests
{
    private static PortableFaultException Fault(Action action) =>
        Assert.Throws<PortableFaultException>(action)!;

    // PB-18-LOCAL-DENIAL-EMITS-NO-FRAME and PB-56-OBSERVATION-COMPLETE-ON-DENIAL
    [Test]
    public async Task A_local_denial_emits_no_frame_and_still_produces_a_complete_observation()
    {
        var handler = new CoolingPortableHandler(CoolingPortableFixture.CreateNativeRegistry());
        await using var host = await PortableTestHarness.DirectHostAsync(handler: handler);
        var result = await host.InvokeAsync(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV1,
            CoolingPortableFixture.Command("primary", enabled: true),
            PortableConstraint.AllOf(
                PortableConstraint.Atom(PortableTruth.Satisfied),
                PortableConstraint.Atom(PortableTruth.Unsatisfied)));

        Assert.Multiple(() =>
        {
            Assert.That(result.FrameDecision, Is.EqualTo(PortableFrameDecision.None));
            Assert.That(result.ResultClass, Is.EqualTo(PortableResultClass.Denial));
            Assert.That(result.Observation.TerminalStatus, Is.EqualTo(PortableTerminalStatus.Denied));
            Assert.That(result.Observation.AuthorityDecision, Is.EqualTo(PortableAuthorityDecision.Denied));
            Assert.That(
                result.Observation.AuthorityDecisionPoint,
                Is.EqualTo(PortableAuthorityDecisionPoint.HostLocal));
            Assert.That(result.Observation.ProviderEffectCount, Is.Zero);
            Assert.That(result.Observation.CopyCount, Is.Zero);
            Assert.That(result.Observation.CrossedBoundaries, Is.EqualTo(new[] { "none" }).AsCollection);
            Assert.That(result.Observation.Interrupted, Is.False);
            Assert.That(result.Observation.RetryCount, Is.Zero);
            Assert.That(handler.ProviderEffectCount, Is.Zero);
        });
    }

    // PB-19-AUTHORITY-VALUE-NOT-PROJECTED
    [Test]
    public async Task An_unrecognized_Constraint_value_version_denies_rather_than_projecting()
    {
        await using var host = await PortableTestHarness.DirectHostAsync();

        // The evaluator recognizes Constraint value version 1 only. Projecting the version 2 value
        // would discard its narrowing semantics and widen authority, so the atom stays Unknown.
        var result = await host.InvokeAsync(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV1,
            CoolingPortableFixture.Command("primary", enabled: true),
            PortableConstraint.AllOf(
                PortableConstraint.Atom(PortableTruth.Satisfied),
                PortableConstraint.Atom(PortableTruth.Unknown)));

        Assert.Multiple(() =>
        {
            Assert.That(result.FrameDecision, Is.EqualTo(PortableFrameDecision.None));
            Assert.That(result.ResultClass, Is.EqualTo(PortableResultClass.Denial));
            Assert.That(result.Observation.AuthorityDecision, Is.EqualTo(PortableAuthorityDecision.Unknown));
            Assert.That(result.Observation.ProviderEffectCount, Is.Zero);
        });
    }

    // PB-20-STRONG-KLEENE-ANYOF-PERMITS and PB-21-STRONG-KLEENE-ALLOF-DENIES
    [Test]
    public void Strong_Kleene_evaluation_permits_only_a_true_expression()
    {
        var satisfied = PortableConstraint.Atom(PortableTruth.Satisfied);
        var unknown = PortableConstraint.Atom(PortableTruth.Unknown);
        var unsatisfied = PortableConstraint.Atom(PortableTruth.Unsatisfied);

        Assert.Multiple(() =>
        {
            Assert.That(PortableConstraint.AnyOf(satisfied, unknown).Evaluate(), Is.EqualTo(PortableTruth.Satisfied));
            Assert.That(PortableConstraint.AllOf(satisfied, unknown).Evaluate(), Is.EqualTo(PortableTruth.Unknown));
            Assert.That(PortableConstraint.AnyOf(unsatisfied, unknown).Evaluate(), Is.EqualTo(PortableTruth.Unknown));
            Assert.That(PortableConstraint.AllOf(satisfied, satisfied).Evaluate(), Is.EqualTo(PortableTruth.Satisfied));
            Assert.That(PortableConstraint.Not(unknown).Evaluate(), Is.EqualTo(PortableTruth.Unknown));
        });
    }

    [Test]
    public async Task An_AnyOf_expression_containing_a_satisfied_atom_may_emit()
    {
        await using var host = await PortableTestHarness.DirectHostAsync();
        var decision = host.PrepareRequest(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV1,
            PortableConstraint.AnyOf(
                PortableConstraint.Atom(PortableTruth.Satisfied),
                PortableConstraint.Atom(PortableTruth.Unknown)));

        Assert.Multiple(() =>
        {
            Assert.That(decision.FrameDecision, Is.EqualTo(PortableFrameDecision.Emit));
            Assert.That(decision.ResultClass, Is.EqualTo(PortableResultClass.Request));
            Assert.That(decision.DenialObservation, Is.Null);
        });
    }

    // PB-22-CAPABILITY-IN-BODY-ACROSS-TRUST
    [Test]
    public async Task A_Capability_in_a_body_across_a_trust_boundary_fails_closed()
    {
        var handler = new CoolingPortableHandler(CoolingPortableFixture.CreateNativeRegistry());
        await using var host = await PortableTestHarness.DirectHostAsync(handler: handler);
        var result = await host.InvokeAsync(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV1,
            CoolingPortableFixture.Command("primary", enabled: true)
                .WithField("capability", new PortableTextValue("cooling.write")),
            PortableTestHarness.Permitted());

        Assert.Multiple(() =>
        {
            Assert.That(result.ResultClass, Is.EqualTo(PortableResultClass.ProtocolError));
            Assert.That(result.Category, Is.EqualTo(PortableProtocolCategory.InvalidAuthorityPresentation));
            Assert.That(handler.ProviderEffectCount, Is.Zero);
        });
    }

    // PB-23-MISSING-NO-CAPABILITY-TRANSFER-DECLARATION
    [Test]
    public void A_trust_crossing_contract_without_the_declaration_does_not_default_to_permissive()
    {
        var permissive = CoolingPortableFixture.Contract with
        {
            Authority = CoolingPortableFixture.Contract.Authority with { NoCapabilityTransfer = false }
        };

        Assert.That(
            Fault(() => PortableNegotiation.Negotiate(
                permissive,
                permissive,
                PortableRealization.NegotiatedProcess,
                "host",
                "provider",
                "test")).Category,
            Is.EqualTo(PortableProtocolCategory.InvalidAuthorityPresentation));
    }

    // PB-24-MISSING-REQUIRED-FRAGMENT
    [Test]
    public async Task A_missing_required_Fragment_is_refused_before_provider_activation()
    {
        var handler = new CoolingPortableHandler(CoolingPortableFixture.CreateNativeRegistry());
        await using var host = await PortableTestHarness.DirectHostAsync(handler: handler);
        var result = await host.InvokeAsync(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV1,
            CoolingPortableFixture.Command("primary", enabled: true, requesterLabel: null),
            PortableTestHarness.Permitted());

        Assert.Multiple(() =>
        {
            Assert.That(result.Category, Is.EqualTo(PortableProtocolCategory.InvalidPayload));
            Assert.That(handler.ProviderEffectCount, Is.Zero, "Attribution is never inferred from delivery.");
        });
    }

    // PB-25-COPIED-IMMUTABLE-BLOB-ACCEPTED
    [Test]
    public async Task A_copied_immutable_blob_is_accepted_and_reports_its_ownership()
    {
        await using var host = await PortableTestHarness.DirectHostAsync();
        var result = await host.InvokeAsync(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV1,
            CoolingPortableFixture.Command("primary", enabled: true),
            PortableTestHarness.Permitted(),
            [PortableTestHarness.Blob()]);

        Assert.Multiple(() =>
        {
            Assert.That(result.ResultClass, Is.EqualTo(PortableResultClass.OutcomeSucceeded));
            Assert.That(result.Observation.ReferencedResources, Has.Length.EqualTo(1));
            Assert.That(result.Observation.ReferencedResources[0].Ownership, Is.EqualTo("copied"));
            Assert.That(result.Observation.ReferencedResources[0].IntegrityVerified, Is.True);
            Assert.That(result.Observation.CopyCount, Is.Zero, "A direct call materializes no copy.");
        });
    }

    // PB-26-RESOURCE-INTEGRITY-MISMATCH
    [Test]
    public void A_content_hash_that_does_not_verify_is_refused_before_any_provider_effect()
    {
        var tampered = new PortableCopiedBlobResource(
            "profile",
            System.Text.Encoding.UTF8.GetBytes("tampered"),
            PortableCopiedBlobResource.HashOf(System.Text.Encoding.UTF8.GetBytes("original")));

        Assert.That(
            Fault(() => PortableResourceCodec.Admit(
                tampered,
                [PortableResourceFlavors.CopiedImmutableBlob],
                [],
                PortableLimits.Declared)).Category,
            Is.EqualTo(PortableProtocolCategory.InvalidPayload));
    }

    // PB-27-ADDRESSING-HANDLE-ACCEPTED
    [Test]
    public async Task An_addressing_only_handle_inside_the_accept_list_crosses_no_octets()
    {
        await using var host = await PortableTestHarness.DirectHostAsync(
            CatalogPortableFixture.Contract,
            new CatalogPortableHandler());
        var result = await host.InvokeAsync(
            CatalogPortableFixture.Upsert,
            CatalogPortableFixture.UpsertCommand,
            CatalogPortableFixture.UpsertCommandValue(CatalogPortableFixture.ItemValue("a", "Alpha", "one")),
            PortableTestHarness.Permitted(),
            [CatalogPortableFixture.Handle()]);

        Assert.Multiple(() =>
        {
            Assert.That(result.ResultClass, Is.EqualTo(PortableResultClass.OutcomeSucceeded));
            Assert.That(result.Observation.ReferencedResources[0].Ownership, Is.EqualTo("provider-retained"));
            Assert.That(result.Observation.CopyCount, Is.Zero);
        });
    }

    // PB-28-RESOURCE-SCOPE-REFUSED
    [Test]
    public void A_handle_outside_the_accept_list_is_a_payload_decision_with_a_non_normative_local_code()
    {
        var fault = Fault(() => PortableResourceCodec.Admit(
            CatalogPortableFixture.Handle("catalog-provider", "secondary"),
            [PortableResourceFlavors.AddressingOnlyHandle],
            [CatalogPortableFixture.AcceptedHandle],
            PortableLimits.Declared));

        Assert.Multiple(() =>
        {
            Assert.That(fault.Category, Is.EqualTo(PortableProtocolCategory.InvalidPayload));
            Assert.That(fault.LocalCode, Is.EqualTo("resource-refused"));
            Assert.That(
                fault.Category,
                Is.Not.EqualTo(PortableProtocolCategory.InvalidAuthorityPresentation),
                "A handle carries no authority, so refusing one is not an authority decision.");
        });
    }

    // PB-29-UNSUPPORTED-RESOURCE-FLAVOR
    [Test]
    public void A_zero_one_non_goal_flavor_fails_negotiation_with_no_silent_downgrade()
    {
        foreach (var nonGoal in PortableResourceFlavors.NonGoals)
        {
            var contract = CoolingPortableFixture.Contract with
            {
                Representation = CoolingPortableFixture.Contract.Representation with { ResourceFlavors = [nonGoal] }
            };

            Assert.That(
                Fault(() => PortableNegotiation.Negotiate(
                    contract,
                    contract,
                    PortableRealization.NegotiatedProcess,
                    "host",
                    "provider",
                    "test")).Category,
                Is.EqualTo(PortableProtocolCategory.UnsupportedContract),
                $"Flavor '{nonGoal}' must fail closed.");
        }
    }

    // PB-30-FORBIDDEN-IMPLICIT-COPY
    [Test]
    public void Resource_octets_alongside_a_handle_are_refused_rather_than_silently_materialized()
    {
        var frame = CborMap.Of(
        [
            ("flavor", new CborText(PortableResourceFlavors.AddressingOnlyHandle)),
            ("name", new CborText("catalog")),
            ("provider", new CborText("catalog-provider")),
            ("id", new CborText("primary")),
            ("content", new CborBytes(new byte[] { 1, 2, 3 }))
        ]);

        var fault = Fault(() => PortableResourceCodec.Decode(
            frame,
            [PortableResourceFlavors.AddressingOnlyHandle],
            [CatalogPortableFixture.AcceptedHandle],
            PortableLimits.Declared));

        Assert.Multiple(() =>
        {
            Assert.That(fault.Category, Is.EqualTo(PortableProtocolCategory.InvalidPayload));
            Assert.That(fault.LocalCode, Is.EqualTo("forbidden-implicit-copy"));
        });
    }

    // PB-31-RESOURCE-EXCEEDS-DECLARED-BOUND
    [Test]
    public void A_resource_beyond_the_declared_bound_is_refused_before_uncontrolled_work()
    {
        var oversized = PortableCopiedBlobResource.Create(
            "profile",
            new byte[PortableLimits.Declared.MaxResourceBytes + 1]);

        Assert.That(
            Fault(() => PortableResourceCodec.Admit(
                oversized,
                [PortableResourceFlavors.CopiedImmutableBlob],
                [],
                PortableLimits.Declared)).Category,
            Is.EqualTo(PortableProtocolCategory.LimitExceeded));
    }

    // PB-32-RELEASE-SIGNAL-FOR-COPIED-FLAVOR
    [Test]
    public void A_release_signal_for_a_flavor_that_defines_none_is_not_a_harmless_no_op()
    {
        var blob = PortableTestHarness.Blob();
        var frame = CborMap.Of(
        [
            ("flavor", new CborText(PortableResourceFlavors.CopiedImmutableBlob)),
            ("name", new CborText(blob.Name)),
            ("content", new CborBytes(blob.Content)),
            ("integrity", new CborText(blob.Integrity)),
            ("release", new CborBoolean(true))
        ]);

        Assert.That(
            Fault(() => PortableResourceCodec.Decode(
                frame,
                [PortableResourceFlavors.CopiedImmutableBlob],
                [],
                PortableLimits.Declared)).Category,
            Is.EqualTo(PortableProtocolCategory.StateViolation));
    }
}

File: Minimal/tests/Brontide.Minimal.Interchange.Tests/Portable/PortableAuthorityAndResourceTests.fs
namespace Brontide.Minimal.Interchange.Tests.Portable

open System
open System.Text
open NUnit.Framework
open Brontide.Minimal.Binding.Portable

/// C3 and C6: local authority, no-capability-transfer, and referenced resources.
[<TestFixture>]
type PortableAuthorityAndResourceTests() =

    let limits = PortableLimits.declared

    let denied = PortableConstraint.Atom PortableTruth.Unsatisfied

    /// An unrecognized Constraint value version makes the atom Unknown; it is never projected onto a
    /// version the evaluator does recognize, because that would widen authority.
    let unrecognizedConstraintVersion = PortableConstraint.Atom PortableTruth.Unknown

    let coolingHostAndHandler () =
        let handler = CoolingHandler()
        directHost CoolingFixture.contract handler, handler

    [<Test>]
    member _.``PB-18 a local denial emits no frame and starts no provider``() =
        let host, handler = coolingHostAndHandler ()

        let result =
            host
                .Invoke(
                    CoolingFixture.setEnabled,
                    CoolingFixture.commandV1,
                    CoolingFixture.authorizedCommand "primary" true,
                    denied
                )
                .Result

        assertAll (fun () ->
            Assert.That(result.FrameDecision, Is.EqualTo FrameDecision.None)
            Assert.That(result.ResultClass, Is.EqualTo ResultClass.Denial)
            Assert.That(result.Observation.ProviderEffectCount, Is.EqualTo(Some 0L))
            Assert.That(handler.ProviderEffectCount, Is.EqualTo 0L))

    [<Test>]
    member _.``PB-19 an unrecognized Constraint value version denies rather than projecting``() =
        let host, handler = coolingHostAndHandler ()

        let result =
            host
                .Invoke(
                    CoolingFixture.setEnabled,
                    CoolingFixture.commandV1,
                    CoolingFixture.authorizedCommand "primary" true,
                    unrecognizedConstraintVersion
                )
                .Result

        assertAll (fun () ->
            Assert.That(result.FrameDecision, Is.EqualTo FrameDecision.None)
            Assert.That(result.ResultClass, Is.EqualTo ResultClass.Denial)
            Assert.That(result.Observation.AuthorityDecision, Is.EqualTo AuthorityDecision.Unknown)
            Assert.That(handler.ProviderEffectCount, Is.EqualTo 0L))

    [<Test>]
    member _.``PB-20 AnyOf with a satisfied atom permits despite an unknown one``() =
        let host = directCoolingHost ()

        let decision =
            expectOk (
                host.PrepareRequest(
                    CoolingFixture.setEnabled,
                    CoolingFixture.commandV1,
                    PortableConstraint.AnyOf
                        [ PortableConstraint.Atom PortableTruth.Satisfied
                          PortableConstraint.Atom PortableTruth.Unknown ]
                )
            )

        assertAll (fun () ->
            Assert.That(decision.FrameDecision, Is.EqualTo FrameDecision.Emit)
            Assert.That(decision.ResultClass, Is.EqualTo ResultClass.Request)
            Assert.That(decision.Admission.Decision, Is.EqualTo AuthorityDecision.Permitted))

    [<Test>]
    member _.``PB-21 AllOf with an unknown atom denies without a far-side effect``() =
        let host, handler = coolingHostAndHandler ()

        let result =
            host
                .Invoke(
                    CoolingFixture.setEnabled,
                    CoolingFixture.commandV1,
                    CoolingFixture.authorizedCommand "primary" true,
                    PortableConstraint.AllOf
                        [ PortableConstraint.Atom PortableTruth.Satisfied
                          PortableConstraint.Atom PortableTruth.Unknown ]
                )
                .Result

        assertAll (fun () ->
            Assert.That(result.FrameDecision, Is.EqualTo FrameDecision.None)
            Assert.That(result.ResultClass, Is.EqualTo ResultClass.Denial)
            Assert.That(result.Observation.AuthorityDecision, Is.EqualTo AuthorityDecision.Unknown)
            Assert.That(handler.ProviderEffectCount, Is.EqualTo 0L))

    [<Test>]
    member _.``strong Kleene negation of unknown stays unknown``() =
        Assert.That(
            PortableConstraint.evaluate (PortableConstraint.Not(PortableConstraint.Atom PortableTruth.Unknown)),
            Is.EqualTo PortableTruth.Unknown
        )

    [<Test>]
    member _.``PB-22 a Capability in the body of a trust-crossing binding fails closed``() =
        let host, handler = coolingHostAndHandler ()

        let carrying =
            CoolingFixture.authorizedCommand "primary" true
            |> PortableRecord.withField "capability" (PortableText "cooling.write@1")

        let result = invoke host CoolingFixture.setEnabled CoolingFixture.commandV1 carrying

        assertAll (fun () ->
            Assert.That(result.FrameDecision, Is.EqualTo FrameDecision.Reject)
            Assert.That(result.Category, Is.EqualTo(Some ProtocolCategory.InvalidAuthorityPresentation))
            Assert.That(handler.ProviderEffectCount, Is.EqualTo 0L))

    [<Test>]
    member _.``PB-23 a trust-crossing contract without the declaration fails closed``() =
        let permissive =
            { CoolingFixture.contract with
                Authority =
                    { CoolingFixture.contract.Authority with
                        NoCapabilityTransfer = false } }

        expectCategory
            ProtocolCategory.InvalidAuthorityPresentation
            (PortableNegotiation.negotiate permissive permissive Realization.FixedDirectCall "host" "provider" "fixed")
        |> ignore

    [<Test>]
    member _.``PB-24 a missing required Fragment is refused before provider activation``() =
        let host, handler = coolingHostAndHandler ()
        let unattributed = CoolingFixture.command "primary" true None None None

        let result = invoke host CoolingFixture.setEnabled CoolingFixture.commandV1 unattributed

        assertAll (fun () ->
            Assert.That(result.Category, Is.EqualTo(Some ProtocolCategory.InvalidPayload))
            Assert.That(handler.ProviderEffectCount, Is.EqualTo 0L))

    [<Test>]
    member _.``PB-25 a copied immutable blob is accepted and observed as copied``() =
        let host = directCoolingHost ()

        let result =
            host
                .Invoke(
                    CoolingFixture.setEnabled,
                    CoolingFixture.commandV1,
                    CoolingFixture.authorizedCommand "primary" true,
                    permitted,
                    [ blob "cooling-profile" ]
                )
                .Result

        assertAll (fun () ->
            Assert.That(result.ResultClass, Is.EqualTo ResultClass.OutcomeSucceeded)
            Assert.That(List.length result.Observation.ReferencedResources, Is.EqualTo 1)

            let observed = List.head result.Observation.ReferencedResources
            Assert.That(observed.Ownership, Is.EqualTo "copied")
            Assert.That(observed.IntegrityVerified, Is.True)
            Assert.That(observed.Accepted, Is.True)
            // A direct call materializes no copy at all.
            Assert.That(result.Observation.CopyCount, Is.EqualTo 0L))

    [<Test>]
    member _.``PB-26 a content hash that does not verify is refused before any effect``() =
        let tampered =
            CopiedBlob("profile", Text.Encoding.UTF8.GetBytes "cooling-profile", String.replicate 64 "a")

        expectCategory
            ProtocolCategory.InvalidPayload
            (ResourceCodec.admit tampered [ ResourceFlavor.CopiedImmutableBlobToken ] [] limits)
        |> ignore

    [<Test>]
    member _.``PB-27 a handle inside the accept list conveys addressing and no octets``() =
        let host = directHost CatalogFixture.contract (CatalogHandler())

        let result =
            host
                .Invoke(
                    CatalogFixture.upsert,
                    CatalogFixture.upsertCommand,
                    CatalogFixture.upsertCommandValue [ CatalogFixture.itemValue "a" "Alpha" [ "cold" ] ],
                    permitted,
                    [ CatalogFixture.handle "catalog-provider" "primary" ]
                )
                .Result

        assertAll (fun () ->
            Assert.That(result.ResultClass, Is.EqualTo ResultClass.OutcomeSucceeded)
            Assert.That(result.Observation.CopyCount, Is.EqualTo 0L)

            let observed = List.head result.Observation.ReferencedResources
            Assert.That(observed.Ownership, Is.EqualTo "provider-retained")
            Assert.That(observed.IntegrityVerified, Is.False))

    [<Test>]
    member _.``PB-28 a handle outside the accept list is a payload refusal, not an authority one``() =
        let outside = CatalogFixture.handle "catalog-provider" "secondary"

        let fault =
            expectCategory
                ProtocolCategory.InvalidPayload
                (ResourceCodec.admit
                    outside
                    [ ResourceFlavor.AddressingOnlyHandleToken ]
                    [ CatalogFixture.acceptedHandle ]
                    limits)

        // The local code stays non-normative; the portable category is what decides.
        Assert.That(fault.LocalCode, Is.EqualTo "resource-refused")

    [<Test>]
    member _.``PB-29 a 0.1 non-goal resource flavor fails negotiation closed``() =
        let borrowed =
            { CatalogFixture.contract with
                Representation =
                    { CatalogFixture.contract.Representation with
                        ResourceFlavors = [ "borrowed-read-only-region" ]
                        AcceptedResourceHandles = [] } }

        Assert.That(ResourceFlavor.nonGoals, Contains.Item "borrowed-read-only-region")

        expectCategory
            ProtocolCategory.UnsupportedContract
            (PortableNegotiation.negotiate borrowed borrowed Realization.FixedDirectCall "host" "provider" "fixed")
        |> ignore

    [<Test>]
    member _.``PB-30 resource octets for a handle-flavored resource are a forbidden implicit copy``() =
        let encoded =
            CborMap
                [ "flavor", CborText ResourceFlavor.AddressingOnlyHandleToken
                  "name", CborText "catalog"
                  "provider", CborText "catalog-provider"
                  "id", CborText "primary"
                  "content", CborBytes [| 1uy; 2uy; 3uy |] ]

        let fault =
            expectCategory
                ProtocolCategory.InvalidPayload
                (ResourceCodec.decode
                    encoded
                    [ ResourceFlavor.AddressingOnlyHandleToken ]
                    [ CatalogFixture.acceptedHandle ]
                    limits)

        Assert.That(fault.LocalCode, Is.EqualTo "forbidden-implicit-copy")

    [<Test>]
    member _.``PB-31 a resource beyond the declared bound is refused before uncontrolled work``() =
        let oversized =
            PortableResource.blob "profile" (Array.zeroCreate (limits.MaxResourceBytes + 1))

        expectCategory
            ProtocolCategory.LimitExceeded
            (ResourceCodec.admit oversized [ ResourceFlavor.CopiedImmutableBlobToken ] [] limits)
        |> ignore

    [<Test>]
    member _.``PB-32 a release signal for the copied flavor is a state violation``() =
        let encoded =
            CborMap
                [ "flavor", CborText ResourceFlavor.CopiedImmutableBlobToken
                  "name", CborText "profile"
                  "content", CborBytes(Text.Encoding.UTF8.GetBytes "cooling-profile")
                  "integrity", CborText(PortableResource.hashOf (Text.Encoding.UTF8.GetBytes "cooling-profile"))
                  "release", CborText "completed" ]

        expectCategory
            ProtocolCategory.StateViolation
            (ResourceCodec.decode encoded [ ResourceFlavor.CopiedImmutableBlobToken ] [] limits)
        |> ignore

    [<Test>]
    member _.``PB-56 a denial still produces a complete observation``() =
        let host = directCoolingHost ()

        let result =
            host
                .Invoke(
                    CoolingFixture.setEnabled,
                    CoolingFixture.commandV1,
                    CoolingFixture.authorizedCommand "primary" true,
                    denied
                )
                .Result

        let observation = result.Observation

        assertAll (fun () ->
            Assert.That(Observation.completenessFailures observation, Is.Empty)
            Assert.That(observation.TerminalStatus, Is.EqualTo TerminalStatus.Denied)
            Assert.That(observation.AuthorityDecisionPoint, Is.EqualTo AuthorityDecisionPoint.HostLocal)
            Assert.That(observation.ProviderEffectCount, Is.EqualTo(Some 0L))
            Assert.That(observation.CopyCount, Is.EqualTo 0L)
            shouldEqual [ "none" ] observation.CrossedBoundaries
            Assert.That(observation.Interrupted, Is.False)
            Assert.That(observation.RetryCount, Is.EqualTo 0L)
            // An unobservable value uses its declared absent form rather than being omitted.
            Assert.That(observation.FailureDomain, Is.EqualTo(Some FailureDomain.LocalEndpoint)))
